pkg/kademlia/kademlia.go

```diff
@@ -153,48 +157,60 @@ func (k *Kademlia) Bootstrap(ctx context.Context) error {
 		return nil
 	}
 
+	waitInterval := k.bootstrapBackoffBase
+
 	var errGroup errs.Group
-	var foundOnlineBootstrap bool
-	for i, node := range k.bootstrapNodes {
-		if ctx.Err() != nil {
-			errGroup.Add(ctx.Err())
-			return errGroup.Err()
+	for i := 0; waitInterval < k.bootstrapBackoffMax; i++ {
+		if i > 0 {
+			time.Sleep(waitInterval)
+			waitInterval = waitInterval * 2
 		}
 
-		ident, err := k.dialer.FetchPeerIdentityUnverified(ctx, node.Address.Address)
-		if err != nil {
-			errGroup.Add(err)
+		var foundOnlineBootstrap bool
+		for i, node := range k.bootstrapNodes {
+			if ctx.Err() != nil {
+				errGroup.Add(ctx.Err())
+				return errGroup.Err()
+			}
+
+			ident, err := k.dialer.FetchPeerIdentityUnverified(ctx, node.Address.Address)
+			if err != nil {
+				errGroup.Add(err)
+				continue
+			}
+
+			k.routingTable.mutex.Lock()
+			node.Id = ident.ID
+			k.bootstrapNodes[i] = node
+			k.routingTable.mutex.Unlock()
+			foundOnlineBootstrap = true
+		}
+
+		if !foundOnlineBootstrap {
+			errGroup.Add(Error.New("no bootstrap node found online"))
 			continue
 		}
 
+		//find nodes most similar to self
 		k.routingTable.mutex.Lock()
-		node.Id = ident.ID
-		k.bootstrapNodes[i] = node
+		id := k.routingTable.self.Id
 		k.routingTable.mutex.Unlock()
-		foundOnlineBootstrap = true
-	}
-
-	if !foundOnlineBootstrap {
-		err := errGroup.Err()
+		_, err := k.lookup(ctx, id, true)
 		if err != nil {
-			return err
+			errGroup.Add(err)
+			continue
 		}
+		return nil
+		// TODO(dylan): We do not currently handle this last bit of behavior.
+		// ```
+		// Finally, u refreshes all k-buckets further away than its closest neighbor.
+		// During the refreshes, u both populates its own k-buckets and inserts
+		// itself into other nodes' k-buckets as necessary.
+		// ```
 	}
 
-	//find nodes most similar to self
-	k.routingTable.mutex.Lock()
-	id := k.routingTable.self.Id
-	k.routingTable.mutex.Unlock()
-	_, err := k.lookup(ctx, id, true)
-
-	// TODO(dylan): We do not currently handle this last bit of behavior.
-	// ```
-	// Finally, u refreshes all k-buckets further away than its closest neighbor.
-	// During the refreshes, u both populates its own k-buckets and inserts
-	// itself into other nodes' k-buckets as necessary.
-	// ``
-
-	return err
+	errGroup.Add(Error.New("unable to start bootstrap after final wait time of %ds", waitInterval/time.Second))
```

Comment:
time.Duration automatically formats well as string: https://play.golang.org/p/nMc2lerNcnK. No need to for this division.

Reply:
cool! fixed